### PR TITLE
OMI_audio_emitter

### PR DESCRIPTION
**Pull request by robertlong**
This is a work in progress draft of a glTF 2.0 extension for audio emitters. It is inspired by the Mozilla Hubs `audio-emitter` component of the`MOZ_hubs_components` extension and the `MSFT_audio_emitter` extension. The properties are closely related to the HTML `<audio>` element and Web Audio `AudioNode` and `PannerNode` APIs.

Some differences from the Hubs/Microsoft extensions:
- Multiple global audio emitters can be placed on the scene
- Audio clips are defined outside of the audio emitter objects for reuse in multiple emitters.
- Audio clips can reference multiple audio files for better content negotiation in different browsers/game engines.
- Emitters are not triggerable via animations for simplicity. This may be added in another extension.
- The `autoPlay` property can either play the audio clip on scene load or when set to false, it can be triggered externally.

[Rendered](https://github.com/omigroup/gltf-extensions/tree/5f32b79492d730956028a3de9e0cbf9dc604151b/extensions/2.0/OMI_audio_emitter)

Out of Scope:
- Audio triggers
- Audio effects / graphs

Research On Other Engines:
https://github.com/omigroup/gltf-extensions/issues/2

### CODE REVIEW

**Comment by fire** [comment]
@humbletim

We want to compare multiple implementations of audio features in Game Engines before the meeting.

* Web Audio
* Godot Engine
* Unity
* Unreal Engine
* Wwise
* OpenAL

**Comment by fire** [comment]
The audio emitter uses an attenuation inner cone is not possible in Godot Engine.

Trying to look for a fallback.

**Comment by humbletim** [comment]
Would it make sense to include support for descriptive `alt` tags / `aria` labels as part of audio emitters? (or maybe an accessibility extension on its own?)

**Comment by robertlong** [comment]
> Would it make sense to include support for descriptive `alt` tags / `aria` labels as part of audio emitters? (or maybe an accessibility extension on its own?)

I think it would make more sense to have an accessibility extension that can be applied to any node.

**Comment by robertlong** [comment]
> The audio emitter uses an attenuation inner cone is not possible in Godot Engine.
> 
> Trying to look for a fallback.

Yeah, looking at the [AudioStreamPlayer3D](https://docs.godotengine.org/en/stable/classes/class_audiostreamplayer3d.html#class-audiostreamplayer3d-property-emission-angle-degrees) class it may require some additional code. I'll look into this further next week. Let me know if you have any ideas. I think we'll run into similar issues with Unity.

File: extensions/2.0/OMI_audio_emitter/README.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            },
+            {
+              "mimeType": "audio/wav",
+              "uri": "./environment.wav"
+            },
+            {
+              "mimeType": "audio/mpeg",
+              "uri": "./environment.mp3"
+            }
+          ]
```
Comment (humbletim) [inline extensions/2.0/OMI_audio_emitter/README.md:25]: Could you elaborate a bit on where fallback audio sources fit into Creator / Platform / Tooling workflows?

Comment (humbletim) [reply extensions/2.0/OMI_audio_emitter/README.md:25]: Thinking more about audio fallbacks and given how multiplexed glTFs might be very difficult to QA and "ear test" (across embedded encodings), it seems like Creators might face a situation where it becomes better to maintain _platform-specific glTF exports_ instead (eg: `scene-ogg.glb`, `scene-mp3.glb`, etc.)...

A few ideas to consider that might help mitigate that sort of thing (including for tomorrow's discussion):
- Could we expect glTF _export tools_ to be slightly more advanced on this front?
    - like, adopt a shared convention where there exists a single-source logical sound that automatically becomes N-way encoded as part of the export config and process?
- Should we include additional metadata in the _glTF node_ to make clip identify clearer?
    - as proposed, fallback sources almost start to look like `KHR_materials_variants`, but here must all represent "the same sound"
    - we could included the logical sound's `duration in seconds` and `[min, max] decibels` to help capture that singular intent (and also help tools to verify consistency / catch unintentional mismatches)
- "User-agent spoofing...?"
   - basically, how can we test different fallbacks at glTF load time -- without actually requiring one-off glTFs?

Comment (robertlong) [reply extensions/2.0/OMI_audio_emitter/README.md:25]: When writing the initial spec I drew inspiration from the `<video>`, `<audio>` and `<picture>` HTML elements which support multiple `<source>` elements. This is largely for cross-browser support. Some engines only support `wav` and `ogg` whereas `mpeg` has some of the best cross-browser support.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source

I wasn't thinking about audio quality though. I think you would try to maintain similar bitrates. And yes, I think the different sources should all represent the same audio clip. Ideally tooling will be available to transcode to the different formats. I also think if you're going to have multiple codecs, you should use audio assets referenced via uri or in external buffers. You should ideally only download the content you need for the platform you are on. However, as with glb vs gltf, it'll be up to the content creator to choose the best format for their use-case.

**Comment by takahirox** [comment]
To clarify for understanding `audioClips`, this extension doesn't limit video file formats. So that `mimeType` is always required, unlike glTF 2.0 core `image` which requires `mimeType` only for `bufferView`, to let user agents detect the video file format is supported. Is my understanding correct?

File: extensions/2.0/OMI_audio_emitter/schema/audioClip.schema.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+                "$ref": "audio.schema.json"
+            },
+            "minItems": 1,
+            "description": "One or more sources for the client to negotiate using the provided mimetypes. Clients should process sources in order."
+        },
+        "extensions": { },
+        "extras": { }
+    },
+    "required": [
+        "audio"
```
Comment (takahirox) [inline extensions/2.0/OMI_audio_emitter/schema/audioClip.schema.json:20]: Not `audio` but `sources` because I couldn't find `audio` in `properties`?

File: extensions/2.0/OMI_audio_emitter/schema/glTF.OMI_audio_emitter.schema.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            "type": "object",
+            "$ref": "audioEmitter.schema.json"
+        },
+        "minItems": 1
+    },
+      "extensions": { },
+      "extras": { }
+  },
+  "required": [
+      "audio"
```
Comment (takahirox) [inline extensions/2.0/OMI_audio_emitter/schema/glTF.OMI_audio_emitter.schema.json:27]: `audioClips` and/or `audioEmitters`?

File: extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+              {
+                  "enum": [ "inverse" ],
+                  "description": "An inverse distance model calculating the gain induced by the distance according to: refDistance / (refDistance + rolloffFactor * (Math.max(distance, refDistance) - refDistance))"
+              },
+              {
+                "enum": [ "exponential" ],
+                "description": "An exponential distance model calculating the gain induced by the distance according to: pow((Math.max(distance, refDistance) / refDistance, -rolloffFactor)"
+              },
+              {
+                  "type": "string"
```
Comment (takahirox) [inline extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json:81]: What other distance models do you expect? And do we really want to allow other models which WebAudio API doesn't support? It may be bad for portability.

File: extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                  "$ref": "glTFid.schema.json"
+              }
+          ],
+          "description": "The id of the audio clip referenced by this audio emitter."
+      },
+      "coneInnerAngle": {
+        "type": "number",
+        "minimum": 0,
+        "maximum": 360,
+        "default": 360
```
Comment (takahirox) [inline extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json:50]: In glTF 2.0 core spec, all angles are in radian. Should we follow it?

File: extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+              },
+              {
+                  "type": "string"
+              }
+          ]
+      },
+      "volume": {
+        "type": "number",
+        "minimum": 0,
+        "default": 1
```
Comment (takahirox) [inline extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json:28]: No `maximum` field? And what volume `1` indicates? Loudest possible value? If so, should `maximum` be `1`?

`volume` in `HTMLMediaElement` must be between `0` and `1`.

https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/volume

Comment (fire) [reply extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json:28]: Many engines use decibles as the default unit for audio.

File: extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json (hunk 5 above)
Comment (fire) [reply extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json:50]: I second radians.

File: extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json (hunk 4 above)
Comment (fire) [reply extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json:81]: I second only supporting a physical model.

File: extensions/2.0/OMI_audio_emitter/schema/audioClip.schema.json (hunk 2 above)
Comment (fire) [reply extensions/2.0/OMI_audio_emitter/schema/audioClip.schema.json:20]: People tend to use audio and sink. "Audio" seems strange.

**Comment by robertlong** [comment]
Notes from meeting:
- Use Radians to conform with glTF Spec
- Still need to look into file format types, but leaning towards MPEG if we can get away with it in Three.js, Unity, Unreal, and Godot
- Inverse distance model doesn't support max distance so don't use it in our example
- Document inner/outer cone and cone gain
- Determine what volume is specified as and what unit it should be represented as and how to implement in each engine, for example Godot which uses Db.

File: extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                  "type": "string"
+              }
+          ],
+          "default": "inverse"
+      },
+      "maxDistance": {
+        "type": "number",
+        "minimum": 0,
+        "exclusiveMinimum": true,
+        "default": 10000
```
Comment (robertlong) [inline extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json:90]: Needs a description that makes it clear that you can only use this for linear distance model

**Comment by humbletim** [comment]
2021.08.05 my additional gltf-extensions meeting notes:

* last week @robertlong  completed initial `three-omi` importer/exporter
    - see: https://github.com/omigroup/three-omi
    - enables `OMI_audio_emitter` glTFs to be used across three.js ecosystem
    - (pending) later paired with `omi-gltf-transform` this may also help to reuse Spoke content (independently from Hubs)

* robert planning to create `omi-gltf-transform` command-line tool
    - will enable Hubs => OMI glTFs conversion
    - (maybe) will enable reverse too (OMI => Hubs glTFs)

* robert made significant progress on `omi-blender-gltf` addon
    - see: https://github.com/omigroup/omi-blender-gltf
    - plan is to support adding `OMI_audio_emitter` elements + export to glTF from blender
    - plugs into existing khronos glTF add-on using its official python hooks
    - exported scenes anticipated to be compatible with `three-omi`
    - (later on) anticipated to be compatible with @fire's Godot importer
    - (help needed) robert is planning to circle back next week after `omi-gltf-transform`, but if anyone has time to sooner feel free to jump in (python) and help prototype eg: audio node properties panel

* @fire has prototyped a Godot MozHubs Importer
    - https://github.com/V-Sekai/godot-blue-steel-wolf
    - right now just a proof of concept specific to .glb's with Spoke/Hubs extensions
    - next steps are to adapt for supporting .glb's with `OMI_audio_emitter` extension
    - @rangermauve and @KooIaIa potentially interested in collaborating on this

* reminder: next meeting scheduled for Thursday August 12th 2:30PM PDT <t:1628803800:f>

File: extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json (hunk 5 above)
Comment (robertlong) [reply extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json]: Based on [this discussion about the yfov](https://github.com/KhronosGroup/glTF/issues/1806) and the [punctual spotlight spec](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_lights_punctual/schema/light.spot.schema.json#L20). I think we should go with a maximum and minimum value of `6.283185307179586`. We'll write in the readme/description that the maximum value is `2 * PI`.

This will work just fine in Javascript btw:


```js
Math.PI / 180 * 360 === 6.283185307179586
```

```js
6.283185307179586 * 180 / Math.PI === 360
```

File: extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        "minimum": 0,
+        "default": 1
+      },
+      "name": { },
+      "extensions": { },
+      "extras": { }
+  },
+  "required": [
+      "type",
+      "source"
```
Comment (robertlong) [inline extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json:107]: source should be optional

**Comment by mrmetaverse** [comment]
@robertlong (I have pinged you a lot today, sorry) Do you know what the status is of this? Are we waiting to merge for a vote or something? :D

**Comment by robertlong** [comment]
> @robertlong (I have pinged you a lot today, sorry) Do you know what the status is of this? Are we waiting to merge for a vote or something? :D

After these final changes are in, we need to write a markdown explainer. Including explaining the various audio rolloff algorithims.

We also need two implementations of the spec before it can go to a vote. So far we have three-omi and we've been talking about Unity and Godot implementations.

In parallel we're working on Blender tooling to generate new glTF content with this extension.

Once these are done we can take it to a vote.

File: extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json (hunk 6 above)
Comment (robertlong) [reply extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json:29]: We decided to rename this field to gain and go with a unitless multiplier with a range of 0-Infinity. It is discussed more in detail here: https://github.com/omigroup/gltf-extensions/discussions/14#discussioncomment-1274627

File: extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json (hunk 5 above)
Comment (robertlong) [reply extensions/2.0/OMI_audio_emitter/schema/audioEmitter.schema.json]: I've made this change now. I think it should be fine, but could use a second opinion.

File: extensions/2.0/OMI_audio_emitter/README.md (hunk 1 above)
Comment (robertlong) [reply extensions/2.0/OMI_audio_emitter/README.md]: We decided to make mp3 the default audio format and use extensions to add additional formats down the line. More details here: https://github.com/omigroup/gltf-extensions/discussions/14

File: extensions/2.0/OMI_audio_emitter/README.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        },
+        {
+          "name": "duck emitter",
+          "type": "positional",
+          "gain": 0.8,
+          "loop": false,
+          "autoPlay": false,
+          "source": 1,
+          "coneInnerAngle": 6.283185307179586,
+          "coneOuterAngle": 6.283185307179586,
```
Comment (mikeskydev) [inline extensions/2.0/OMI_audio_emitter/README.md:37]: I noticed that `KHR_LIGHTS_PUNCTUAL` uses `innerConeAngle` and `outerConeAngle` to describe it's spotlight cone. As we have updated the proposed specification already to more closely mirror the way images are added, should we also look to other primary extensions for naming inspiration?

**Comment by lyuma** [comment]
I would like to note that the property
```
autoPlay
```
has rather weak semantics, especially in terms of how a future extension to enable animating its state would be defined.

Instead, I would propose that this property be renamed to
```
playing
```
which would work the same as autoplay but imply stronger semantics of what it means to animate its value to `true` or `false` in a future animation extension.

**Comment by robertlong** [comment]
During our meeting today there were no objections to moving from `autoPlay` to `playing`

**Comment by npolys** [comment]
Worthy of study and harmonization:
There is also an example of a spatialized 3D Web Specification (ISO X3D) that is also compatible with the W3C WebAudio standard: https://web3d.org/documents/specifications/19775-1/V4.0/Part01/components/sound.html
